Deferrable mode for S3ToGCSOperator

moiseenkov opened the pull request:
<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

kaxil commented on generated/provider_dependencies.json:
```diff
@@ -359,6 +359,7 @@
       "google-cloud-secret-manager>=0.2.0,<2.0.0",
       "google-cloud-spanner>=1.10.0,<2.0.0",
       "google-cloud-speech>=0.36.3,<2.0.0",
+      "google-cloud-storage-transfer<=1.4.0",
```
Shouldn't pin unless major version

moiseenkov replied:
Unfortunately, later versions of `google-cloud-storage-transfer` lead dependencies conflict. It can be upgraded together with other google packages. I left a comment about it in the `airflow/providers/google/provider.yaml`.

kaxil replied:
That will just create more issues in the future though. Could `<2` work?

moiseenkov replied:
Unfortunately, `google-cloud-storage-transfer` of version 1.4.1+ requires higher versions of `protobuf` and `proto-plus` libraries that conflict with other Google packages. However, all these packages should be upgraded soon and `google-cloud-storage-transfer` will be among them.

potiuk replied:
We should add a comment about it

kaxil commented on airflow/providers/google/cloud/triggers/cloud_storage_transfer_service.py:
```diff
@@ -0,0 +1,122 @@
+    """
+
+    def __init__(
+        self, job_names: list[str], project_id: str | None = None, polling_interval_seconds: int = 10
```
`polling_interval_seconds` -- should we sync with others and have `poll_interval` instead?

moiseenkov replied:
Thanks for the advice. The name is updated.

kaxil commented on airflow/providers/google/cloud/transfers/s3_to_gcs.py:
```diff
@@ -104,6 +127,7 @@ class S3ToGCSOperator(S3ListOperator):
         "google_impersonation_chain",
     )
     ui_color = "#e09411"
+    transfer_job_max_files_number = 1000
```
Why hard-code? should we make it configurable

moiseenkov replied:
Thank you for noticing. It's a good point, but I think it's better to leave it hard-coded, because:
1. This attribute exist only because of Google Storage Transfer Service API limitation for the field `includePrefixes[]` (docs: https://cloud.google.com/storage-transfer/docs/reference/rest/v1/TransferSpec#ObjectConditions). We can't increase it, because it might cause "400 Bad request" responses from the API, and decreasing it is also not a good idea, because it might increase the number of API calls. That's why I don't see a use case for changing it, unless the API updates.
2. Moving this attribute into the `S3ToGCSOperator`'s parameters list also doesn't look good to me, because it is a part of a low level logic that is out of the operator's scope. This new parameter won't influence the operator's result and thus will be quite useless for users. Moreover, after creating a new parameter we will have to maintain it, and it won't be easy to remove it in the future.

That's why from my perspective exposing it to the users won't bring any value. However, it is always possible to monkey patch it if needed. WDYT?

kaxil commented on docs/apache-airflow-providers-google/operators/transfer/s3_to_gcs.rst:
```diff
@@ -37,6 +37,16 @@ to transfer data from Amazon S3 to Google Cloud Storage.
 
+There is a possibility to start S3ToGCSOperator asynchronously using deferrable mode. To do so just add parameter
+``deferrable=True`` into the operator call. Under the hood it will delegate data transfer to Google Cloud Storage
+Transfer Service. By changing parameter ``polling_interval_seconds=10`` you can control frequency of polling a transfer
```
```suggestion
Transfer Service. By changing parameter ``polling_interval=10`` you can control frequency of polling a transfer
```

moiseenkov replied:
Thanks for noticing, I updated the PR.

kaxil reviewed (commented):
cc @pankajastro Could you review this too please

pankajastro commented on airflow/providers/google/cloud/hooks/cloud_storage_transfer_service.py:
```diff
@@ -490,3 +494,54 @@ def operations_contain_expected_statuses(
+                raise AirflowException(ex)
+        return self._client
+
+    def get_jobs(self, job_names: list[str]):
```
can we make this async fun

moiseenkov replied:
Sure, I also updated  unit tests for this.

pankajastro commented on airflow/providers/google/cloud/transfers/s3_to_gcs.py:
```diff
@@ -184,34 +212,127 @@ def execute(self, context: Context):
+
+    def transfer_files_async(self, files: list[str], gcs_hook: GCSHook, s3_hook: S3Hook):
+        """Submits Google Cloud Storage Transfer Service job to copy files from AWS S3 to GCS"""
+        if self.poll_interval <= 0:
```
shall we check this in transfer operator init only that way it would fail early

moiseenkov replied:
Good point! Fixed.

pankajastro commented on airflow/providers/google/cloud/transfers/s3_to_gcs.py:
```diff
@@ -184,34 +212,127 @@ def execute(self, context: Context):
+        self.log.info("All done, uploaded %d files to Google Cloud Storage", len(files))
+
+    def transfer_files_async(self, files: list[str], gcs_hook: GCSHook, s3_hook: S3Hook):
+        """Submits Google Cloud Storage Transfer Service job to copy files from AWS S3 to GCS"""
```
```suggestion
        """Submits Google Cloud Storage Transfer Service job to copy files from AWS S3 to GCS"""
        
```

pankajastro commented on airflow/providers/google/cloud/triggers/cloud_storage_transfer_service.py:
```diff
@@ -0,0 +1,120 @@
+        async_hook = self.get_async_hook()
+
+        while True:
+            self.log.info("Attempting to request jobs statuses")
```
Adding more meaningful log messages here will help, something like maybe how many job done

moiseenkov replied:
Well, there is such a log message that says how many jobs are completed just a few lines below after the try..except block.

pankajastro commented on tests/providers/google/cloud/hooks/test_cloud_storage_transfer_service_async.py:
```diff
@@ -0,0 +1,162 @@
+        yield CloudDataTransferServiceAsyncHook()
+
+
+if sys.version_info < (3, 8):
```
Let's reuse it from https://github.com/apache/airflow/blob/main/tests/providers/google/cloud/utils/compat.py

moiseenkov replied:
Thanks for noticing. Done.

pankajastro commented on tests/providers/google/cloud/hooks/test_cloud_storage_transfer_service_async.py:
```diff
@@ -0,0 +1,162 @@
+)
+from tests.providers.google.cloud.utils.base_gcp_mock import mock_base_gcp_hook_default_project_id
+
+if sys.version_info < (3, 8):
```
reuse it from https://github.com/apache/airflow/blob/main/tests/providers/google/cloud/utils/compat.py

moiseenkov replied:
Done.

pankajastro commented on airflow/providers/google/cloud/transfers/s3_to_gcs.py:
```diff
@@ -184,34 +212,127 @@ def execute(self, context: Context):
+                AWS_S3_DATA_SOURCE: {
+                    BUCKET_NAME: self.bucket,
+                    AWS_ACCESS_KEY: {
+                        ACCESS_KEY_ID: config.aws_access_key_id,
```
keep these keys consistent i.e either use import constant for both key or hardcoded one like "secret_access_key"

pankajastro replied:
what if roleArn is givien ?

pankajastro replied:
shall we path param also `path`, WDYT?

moiseenkov replied:
Speaking about consistency, agree with you, thanks for noticing - I fixed that.
Regarding other questions I'm not sure that I understand what exactly you mean. Could you please describe it in more detail?

pankajastro replied:
hmm. I was talking about other aws params https://cloud.google.com/storage-transfer/docs/reference/rest/v1/TransferSpec#AwsS3Data

pankajastro commented on airflow/providers/google/cloud/transfers/s3_to_gcs.py:
```diff
@@ -184,34 +212,127 @@ def execute(self, context: Context):
+        for i in range(0, len(files), chunk_size):
+            files_chunk = files[i : i + chunk_size]
+            body[TRANSFER_SPEC][OBJECT_CONDITIONS]["include_prefixes"] = files_chunk
+            job = transfer_hook.create_transfer_job(body=body)
```
If a job fails in the middle, do we need to cleanup?

moiseenkov replied:
As I see it, it's better to not perform cleaning up or undo operations if the job fails, because current implementation is transparent for users - they are able to see, where the transfer got interrupted, what files were copied and what files still waiting, they can find the failed job and examine it get a better understanding of what happened.

pankajastro commented on airflow/providers/google/cloud/triggers/cloud_storage_transfer_service.py:
```diff
@@ -0,0 +1,120 @@
+
+                for job, operation in zip(jobs, operations):
+                    if operation is None:
+                        yield TriggerEvent(
```
once it fails what will happen to job that you have created in `create_transfer_job`

moiseenkov replied:
If one of the jobs fails then the trigger yields TriggerEvent with information about this failure. After this event handled the DAG run marks failed. Each created transfer job remains in the Google Cloud Storage Service. All jobs work independently thus failed jobs won't interfere with healthy jobs.

pankajastro commented on tests/providers/google/cloud/hooks/test_cloud_storage_transfer_service_async.py:
```diff
@@ -0,0 +1,162 @@
+
+class TestCloudDataTransferServiceAsyncHook:
+    @mock.patch(
+        "airflow.providers.google.cloud.hooks.cloud_storage_transfer_service"
```
This patch path is really long maybe you can keep a constant
```
base_path = airflow.providers.google.cloud.hooks.cloud_storage_transfer_service

@mock.patch(f"{base_path}.storage_transfer_v1.StorageTransferServiceAsyncClient")
def test_(...):
   ...
```

moiseenkov replied:
Agree, thanks for noticing, fixed.

pankajastro commented on airflow/providers/google/cloud/triggers/cloud_storage_transfer_service.py:
```diff
@@ -0,0 +1,119 @@
+    :param project_id: GCP project id.
+    """
+
+    def __init__(self, job_names: list[str], project_id: str | None = None, poll_interval: int = 10):
```
poll_interval missing in docs string

pankajastro commented on airflow/providers/google/cloud/triggers/cloud_storage_transfer_service.py:
```diff
@@ -0,0 +1,119 @@
+    def serialize(self) -> tuple[str, dict[str, Any]]:
+        """Serializes StorageTransferJobsTrigger arguments and classpath."""
+        return (
+            "airflow.providers.google.cloud.triggers.cloud_storage_transfer_service."
```
this string looks long maybe we should do 
`f"{self.__class__.__module__ } .{self.__class__.__qualname__}"`

ahidalgob commented on docs/apache-airflow-providers-google/operators/transfer/s3_to_gcs.rst:
```diff
@@ -37,6 +37,16 @@ to transfer data from Amazon S3 to Google Cloud Storage.
+
+.. exampleinclude::/../tests/system/providers/google/cloud/gcs/example_s3_to_gcs_async.py
+    :language: python
+    :start-after: [START howto_transfer_s3togcs_operator]
```
I think we need to use a different tag in here, don't we? `howto_transfer_s3togcs_operator_async` and also update `example_s3_to_gcs_async.py`

pankajastro commented on airflow/providers/google/cloud/triggers/cloud_storage_transfer_service.py:
```diff
@@ -0,0 +1,119 @@
+class CloudStorageTransferServiceCreateJobsTrigger(BaseTrigger):
+    """
+    StorageTransferJobTrigger run on the trigger worker to perform Cloud Storage Transfer job
+    :param job_names: List of transfer jobs names
```
```suggestion

    :param job_names: List of transfer jobs names
```
This applies at other places also. please keep an empty line between description and param it looks better in UI.

pankajastro commented on airflow/providers/google/cloud/triggers/cloud_storage_transfer_service.py:
```diff
@@ -0,0 +1,119 @@
+
+            jobs_total = len(self.job_names)
+            self.log.info("Transfer jobs completed: %s of %s", jobs_completed_successfully, jobs_total)
+            if jobs_completed_successfully == jobs_total:
```
Just want to point out that the value of jobs_total and jobs_completed_successfully can be invalidated if a trigger die and another trigger picks the job

moiseenkov committed:
Combine 8 into 1

Implement deferrable mode for S3ToGCSOperator

Update airflow/providers/google/cloud/hooks/cloud_storage_transfer_service.py

Co-authored-by: Wei Lee <weilee.rx@gmail.com>

Update airflow/providers/google/cloud/hooks/cloud_storage_transfer_service.py

Co-authored-by: Wei Lee <weilee.rx@gmail.com>

Update airflow/providers/google/cloud/transfers/s3_to_gcs.py

Co-authored-by: Wei Lee <weilee.rx@gmail.com>

Update airflow/providers/google/cloud/transfers/s3_to_gcs.py

Co-authored-by: Wei Lee <weilee.rx@gmail.com>

Update airflow/providers/google/cloud/transfers/s3_to_gcs.py

Co-authored-by: Wei Lee <weilee.rx@gmail.com>

Update airflow/providers/google/cloud/triggers/cloud_storage_transfer_service.py

Co-authored-by: Wei Lee <weilee.rx@gmail.com>

Address code review comments

potiuk reviewed (changes requested):
We had a recent change on how we treat default deferrable setting and we need to adjust this one too: https://github.com/apache/airflow/pull/31712

One more case in point @Lee-W  that we need to add a pre-commit checking the defferrable convention, otherwise we will end up will all kind of mess where some deferrable parameters will have the default values following the convention and some not. 

@moiseenkov   - can you please update the change to follow this patterm? 
@Lee-W -> will it be possible that you add such pre-commit  (https://github.com/apache/airflow/pull/32355#issuecomment-1621100211) ?

Lee-W commented on airflow/providers/google/cloud/hooks/cloud_storage_transfer_service.py:
```diff
@@ -466,3 +472,53 @@ def operations_contain_expected_statuses(
+class CloudDataTransferServiceAsyncHook(GoogleBaseAsyncHook):
+    """Asynchronous hook for Google Storage Transfer Service."""
+
+    def __init__(self, project_id: str | None = None, **kwargs: Any):
```
nitpick
```suggestion
    def __init__(self, project_id: str | None = None, **kwargs: Any) -> None:
```

Lee-W commented on airflow/providers/google/cloud/hooks/cloud_storage_transfer_service.py:
```diff
@@ -466,3 +472,53 @@ def operations_contain_expected_statuses(
+        self.project_id = project_id
+        self._client: storage_transfer_v1.StorageTransferServiceAsyncClient | None = None
+
+    def get_conn(self):
```
nitpick: could we please add a return type annotation here

Lee-W commented on airflow/providers/google/cloud/hooks/cloud_storage_transfer_service.py:
```diff
@@ -466,3 +472,53 @@ def operations_contain_expected_statuses(
+        if not self._client:
+            try:
+                self._client = storage_transfer_v1.StorageTransferServiceAsyncClient()
+            except GoogleAuthError as ex:
```
Do we need to catch the exception here? Should we handle the exception in where this is called? It might be a bit confusing when I call this function and get an `AirflowException` instead of connection related one

Lee-W commented on airflow/providers/google/cloud/hooks/cloud_storage_transfer_service.py:
```diff
@@ -466,3 +472,53 @@ def operations_contain_expected_statuses(
+                raise AirflowException(ex)
+        return self._client
+
+    async def get_jobs(self, job_names: list[str]):
```
nitpick: could we please add a return type annotation here

Lee-W commented on airflow/providers/google/cloud/hooks/cloud_storage_transfer_service.py:
```diff
@@ -466,3 +472,53 @@ def operations_contain_expected_statuses(
+            response_operation = await client.transport.operations_client.get_operation(latest_operation_name)
+            operation = storage_transfer_v1.TransferOperation.deserialize(response_operation.metadata.value)
+            return operation
+        return None
```
I think we don't need it for returning `None`.

```suggestion
```

Lee-W commented on airflow/providers/google/cloud/transfers/s3_to_gcs.py:
```diff
@@ -221,3 +246,104 @@ def gcs_to_s3_object(self, gcs_object: str) -> str:
             return self.prefix + s3_object
         return s3_object
+
+    def transfer_files(self, s3_objects: list[str], gcs_hook: GCSHook, s3_hook: S3Hook):
```
nitpick
```suggestion
    def transfer_files(self, s3_objects: list[str], gcs_hook: GCSHook, s3_hook: S3Hook) -> None:
```

Lee-W commented on airflow/providers/google/cloud/transfers/s3_to_gcs.py:
```diff
@@ -221,3 +246,104 @@ def gcs_to_s3_object(self, gcs_object: str) -> str:
+
+            self.log.info("All done, uploaded %d files to Google Cloud Storage", len(s3_objects))
+
+    def transfer_files_async(self, files: list[str], gcs_hook: GCSHook, s3_hook: S3Hook):
```
nitpick

```suggestion
    def transfer_files_async(self, files: list[str], gcs_hook: GCSHook, s3_hook: S3Hook) -> None:
```

Lee-W commented on airflow/providers/google/cloud/transfers/s3_to_gcs.py:
```diff
@@ -221,3 +246,104 @@ def gcs_to_s3_object(self, gcs_object: str) -> str:
+
+    def transfer_files_async(self, files: list[str], gcs_hook: GCSHook, s3_hook: S3Hook):
+        """Submits Google Cloud Storage Transfer Service job to copy files from AWS S3 to GCS."""
+        if len(files) <= 0:
```
Should we just go with an empty check?

```suggestion
        if not len(files):
```

Lee-W commented on airflow/providers/google/cloud/triggers/cloud_storage_transfer_service.py:
```diff
@@ -0,0 +1,120 @@
+    :param poll_interval: Interval in seconds between polls.
+    """
+
+    def __init__(self, job_names: list[str], project_id: str | None = None, poll_interval: int = 10):
```
nitpick

```suggestion
    def __init__(self, job_names: list[str], project_id: str | None = None, poll_interval: int = 10) -> None:
```

Lee-W commented on tests/system/providers/google/cloud/gcs/example_s3_to_gcs_async.py:
```diff
@@ -0,0 +1,102 @@
+        bucket_name=BUCKET_NAME,
+        project_id=GCP_PROJECT_ID,
+    )
+    # [START howto_transfer_s3togcs_operator_async]
```
```suggestion

    # [START howto_transfer_s3togcs_operator_async]
```
